lib/sqs.js

```diff
@@ -44,6 +76,49 @@ module.exports = ({ account, queuePrefix = '', queueSuffix = '', defaultVisibili
      */
     remove: ({ queueName, entries }) => SQS.deleteMessageBatchAsync({ Entries: _.uniqBy(entries, 'Id'), QueueUrl: getQueueURL(queueName) }),
 
+    /**
+     * Retrieves an extended message to a queue, meaning: if the message is
+     * too large to fit in sqs, it will pull an s3 bucket object if the message
+     * was too large to fit in sqs.
+     * 
+     * @param queueName Name of queue
+     * @param max max number of messages to pull, defaults to 10
+     */
+    extendedRetrieve: ({ queueName, max = 10, messageAttributeNames = [], visibilityTimeout }) => {
+      return SQS.receiveMessageAsync({
+        MaxNumberOfMessages: max,
+        QueueUrl: getQueueURL(queueName),
+        WaitTimeSeconds: 20,
+        VisibilityTimeout: visibilityTimeout || defaultVisibilityTimeout,
+        MessageAttributeNames: messageAttributeNames
+      })
+        .then((messages) => {
+          if (_.isEmpty(messages.Messages)) {
+            return [];
+          }
+
+          return Promise.map(messages.Messages, (message) => {            
+            return decompress(Buffer.from(message.Body, 'base64'))
+              .then((messageBody) => {
+                messageBody = JSON.parse(messageBody);
+                            
+                if (!_.get(messageBody, 'spAwsMetadata._key')) {
+                  return [message, messageBody];
+                }
+
+                return S3.getObjectAsync({
+                  Bucket: messageBody.spAwsMetadata._bucket,
+                  Key: messageBody.spAwsMetadata._key 
+                })
+                  .then((message) => decompress(Buffer.from(message.Body, 'base64')))
+                  .then((message) => JSON.parse(message))
+                  .then((body) => [message, body])
+                  .catch((err) => { return { message, error: err }});
```

Comment:
do we do this in other places in this library?  trying to even wrap my head around what this would look like for the caller, they'd get an array like so if one of them failed?:

`[ [msg, body], [msg, body], {msg, error}, ... `

that feels a little weird to me, maybe b/c I haven't looked at the caller yet

Reply:
no yeah this is definitely weird.  probably should just do objects in both places, i threw the array piece in because i realized we needed the original message object in order to delete from the queue